python/singa/layer.py

```diff
@@ -158,6 +217,21 @@ def _get_unique_name(self):
             self.name = ''
         return self.name
```

Comment:
would it better to create the name in the parent layer?
```python
self.conv = Conv2d()... # when this called, we set self.conv.name = self.name + Layer.sep + 'conv' in _set_attr_

def __set_attr__(self, name, value):
     if isinstance(value, layer):
        ....
       if self.name != "":
         prefix = self.name + Layer.sep
       else:
          prefix = ""
   
        value.name = prefix +  name
     ...
```

Reply:
similarly, we can set the name of the params as:
```python
self.W= Tensor()... # when this called, we set self.W.name = self.name + Layer.ParamSep + 'W' in _set_attr_

def __set_attr__(self, name, value):
     if isinstance(value, Tensor):
        ....
       if self.name != "":
         prefix = self.name + Layer.ParamSep
       else:
          prefix = ""
   
        value.name = prefix +  name
     ...

```

Reply:
1. If _get_unqiue_name is called twice, tha layer name will have duplicateed substrings.
2. I think we can't implement like this, we need to set the name after calling the __init__. If there is a multi-level structure(model->layer1->layer2), the order of function calls is as follows
* model.__init__
* layer1.__init__
* layer2.__init__
* layer1.__setattr__: the global name of layer1 hasn't been set at this time(just a local name), so we can't set the name of layer2.
* model.__setattr__
3. I have thought about this, but we may set names for non-parameter tensors. If users set a name for a tensor that is not a parameter, this method will overwrite the name.

Reply:
2. 
```python
def __set_attr__(self, name, value):
     if isinstance(value, layer) and layer.name is None: # set the name of a layer attribute
             if self.name != "":
                prefix = self.name + Layer.sep
             else:
                prefix = ""
   
              value.name = prefix +  name
     elif name == 'name':  # reset the name of a layer attribute; triggered by the line above.
           for sublayer in self.sublayers:
             sublayer.name = value + Layer.sep + sublayer.name.split(Layer.sep)[-1]
           for pname, pval in self.get_params():
             pval.name = value + Param.sep + pval.name.split(Param.sep)[-1]
```

3. we keep the user assigned name for tensors and do nothing for such cases.

Reply:
2. This scheme will reset the layer name many times. 
```python
1. model.\_\_init\_\_
2. layer1.\_\_init\_\_
3. layer2.\_\_init\_\_
4. layer3.\_\_init\_\_
5. layer2.\_\_setattr\_\_: reset the name of layer3
6. layer1.\_\_setattr\_\_: reset the names of layer2 and layer3
7. model.\_\_setattr\_\_: reset the names of laye1, layer2 and layer3
```
3. the tensor has a default name if user doesn't set the name.https://github.com/apache/singa/blob/master/python/singa/tensor.py#L116

Reply:
2. yes. it will set and then reset the names many times if the depths is large. the advantage is that whenever the parent layer name is changed,all sub layers will be updated. I am fine with both schemes. just need to add some docstring to state that the layer name should not change/reset at runtime.  
3. then just check based on the pattern? if it is like dummy+integer, then it means the name is not set by users.